frontend/apps/reader/modules/readertoc.lua

```diff
@@ -217,11 +217,25 @@ end
 
 function ReaderToc:getAccurateTocIndexByXPointer(xptr)
     local index = self:getTocIndexByPage(xptr)
-    if self.toc[index + 1] then
-        local next_index_xptr = self.toc[index + 1].xpointer
-        local comparison = self.ui.document:compareXPointers(next_index_xptr, xptr)
-        if comparison and comparison > 0 then
-            return index + 1
+    local i = 0
+    local initial_comparison = self.ui.document:compareXPointers(self.toc[index].xpointer, xptr)
+    if initial_comparison and initial_comparison < 0 then
+        while self.toc[index - i] do
+            local toc_xptr = self.toc[index - i].xpointer
+            local cmp = self.ui.document:compareXPointers(toc_xptr, xptr)
+            if cmp and cmp >= 0 then
+                return index - i
+            end
+            i = i + 1
+        end
+    else
+        while self.toc[index + i] do
+            local toc_xptr = self.toc[index + i].xpointer
+            local cmp = self.ui.document:compareXPointers(toc_xptr, xptr)
+            if cmp and cmp <= 0 then
+                return index + i
```

Comment:
Initial comparison: xptr comes after the toc xpointer - so there (in the `else`), you go look at the ones after.
Here, when `cmp <= 0`, you just found a `toc_xptr` that is after `xptr`, so, not a good one (but it is when `cmp == 0`) - but the one just before was allright. So, shouldn't you return `index + i - 1` ?
Overall, logic seems allright, but these `index + i` make that a bit harder to read than it should :) Can you try just using:
```lua
local i = index
self.toc[i]
return i
i = i + 1
```

Reply:
>So, shouldn't you return index + i - 1 ?
Of course it should! I must have missed it, because `cmp <=  0` should be `cmp < 0`. Actually it probably doesn't matter if `getTocIndexByPage` doesn't change. For now it always returns the last "chapter" on the page, thus making `else` unnecessary.